File: durga-rust/src/plugins/json_schema_validator.rs
//! Port of `org.gautelis.durga.plugins.JsonSchemaValidator` (plugin id
//! `json-schema-validator`).
//!
//! Two config forms:
//! * compact `required=a,b,c` (does not start with `{`) — presence check,
//! * a JSON-Schema-subset object — type / required / properties / enum /
//!   minimum / maximum / minLength / maxLength / pattern / items.
//!
//! An optional `onInvalid` directive selects how invalid payloads are handled:
//! `dlq` (default) routes them to the dead-letter channel, `skip` drops them
//! silently, and `fail` fails the process. It is appended alongside the schema,
//! e.g. `schema={"type":"object"};onInvalid=skip` or, in compact form,
//! `required=order_id,amount;onInvalid=fail`.
//!
//! On success the payload is returned unchanged; on failure the plugin returns
//! a [`PluginResult`] carrying the selected error strategy (the generated worker
//! turns that into the corresponding lifecycle event).

use regex::Regex;
use serde_json::Value;

use crate::pipeline::field_at;
use crate::plugin::{Plugin, PluginError};
use crate::result::PluginResult;

#[derive(Default)]
pub struct JsonSchemaValidator;

/// How the plugin handles a payload that fails validation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum OnInvalid {
    Dlq,
    Skip,
    Fail,
}

impl JsonSchemaValidator {
    pub fn new() -> Self {
        JsonSchemaValidator
    }

    pub fn validate(payload: &str, config: &str) -> Result<String, PluginError> {
        let (schema_config, _) = parse_config(config);
        match validate_payload(payload, &schema_config)? {
            Some(err) => Err(err.into()),
            None => Ok(payload.to_string()),
        }
    }
}

/// Separates an optional `onInvalid` directive from the schema configuration.
/// Segments are split on top-level `;` only, so a `;` inside the JSON schema
/// (within braces or string literals) does not act as a separator.
fn parse_config(config: &str) -> (String, OnInvalid) {
    let mut on_invalid = OnInvalid::Dlq;
    let mut schema_config = String::new();
    for segment in split_top_level(config) {
        let trimmed = segment.trim();
        if trimmed.is_empty() {
            continue;
        }
        let lower = trimmed.to_ascii_lowercase();
        if let Some(rest) = lower.strip_prefix("oninvalid=") {
            on_invalid = match rest.trim() {
                "skip" => OnInvalid::Skip,
                "fail" => OnInvalid::Fail,
                _ => OnInvalid::Dlq,
            };
        } else if lower.starts_with("schema=") {
            append_schema(&mut schema_config, &trimmed["schema=".len()..]);
        } else {
            append_schema(&mut schema_config, trimmed);
        }
    }
    (schema_config, on_invalid)
}

fn append_schema(schema_config: &mut String, segment: &str) {
    let value = segment.trim();
    if value.is_empty() {
        return;
    }
    if !schema_config.is_empty() {
        schema_config.push(';');
    }
    schema_config.push_str(value);
}

fn split_top_level(config: &str) -> Vec<String> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut depth = 0i32;
    let mut in_string = false;
    let mut escaped = false;
    for c in config.chars() {
        if in_string {
            current.push(c);
            if escaped {
                escaped = false;
            } else if c == '\\' {
                escaped = true;
            } else if c == '"' {
                in_string = false;
            }
            continue;
        }
        match c {
            '"' => {
                in_string = true;
                current.push(c);
            }
            '{' | '[' => {
                depth += 1;
                current.push(c);
            }
            '}' | ']' => {
                if depth > 0 {
                    depth -= 1;
                }
                current.push(c);
            }
            ';' if depth == 0 => {
                parts.push(std::mem::take(&mut current));
            }
            _ => current.push(c),
        }
    }
    parts.push(current);
    parts
}

fn validate_payload(payload: &str, schema_config: &str) -> Result<Option<String>, PluginError> {
    let input: Value = serde_json::from_str(payload)
        .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;

    let trimmed = schema_config.trim();
    if !trimmed.is_empty() && !trimmed.starts_with('{') {
        return Ok(validate_compact(&input, schema_config));
    }

    let schema: Value = serde_json::from_str(schema_config)
        .map_err(|e| -> PluginError { format!("Invalid schema config: {e}").into() })?;
    Ok(validate_node(&input, &schema, "$"))
}

fn validate_compact(input: &Value, config: &str) -> Option<String> {
    for part in config.split(';') {
        let Some(eq) = part.find('=') else { continue };
        if eq == 0 {
            continue;
        }
        let key = part[..eq].trim();
        let value = part[eq + 1..].trim();
        if key == "required" {
            for field in value.split(',') {
                let path = field.trim();
                if !path.is_empty() && field_at(input, path).is_none() {
                    return Some(format!("$: missing required field '{path}'"));
                }
            }
        }
    }
    None
}

fn validate_node(node: &Value, schema: &Value, path: &str) -> Option<String> {
    let ty = schema.get("type").and_then(|t| t.as_str());
    if let Some(ty) = ty {
        if let Some(err) = check_type(node, ty, path) {
            return Some(err);
        }
    }

    if let Some(allowed) = schema.get("enum") {
        if let Some(arr) = allowed.as_array() {
            if !arr.iter().any(|v| v == node) {
                return Some(format!("{path}: value not in allowed enum"));
            }
        }
    }

    if ty == Some("object") && node.is_object() {
        if let Some(req) = schema.get("required").and_then(|r| r.as_array()) {
            for field in req {
                if let Some(name) = field.as_str() {
                    if node.get(name).is_none() {
                        return Some(format!("{path}: missing required field '{name}'"));
                    }
                }
            }
        }
        if let Some(props) = schema.get("properties").and_then(|p| p.as_object()) {
            for (field, child_schema) in props {
                if let Some(child) = node.get(field) {
                    let child_path = format!("{path}.{field}");
                    if let Some(err) = validate_node(child, child_schema, &child_path) {
                        return Some(err);
                    }
                }
            }
        }
    }

    if node.is_number() {
        if let Some(min) = schema.get("minimum").and_then(|m| m.as_f64()) {
            if node.as_f64().unwrap_or(f64::NAN) < min {
                return Some(format!("{path}: numeric value below minimum {min}"));
            }
        }
        if let Some(max) = schema.get("maximum").and_then(|m| m.as_f64()) {
            if node.as_f64().unwrap_or(f64::NAN) > max {
                return Some(format!("{path}: numeric value above maximum {max}"));
            }
        }
    }

    if let Some(text) = node.as_str() {
        if let Some(min) = schema.get("minLength").and_then(|m| m.as_u64()) {
            if (text.chars().count() as u64) < min {
                return Some(format!("{path}: string length {} below minLength {min}", text.chars().count()));
            }
        }
        if let Some(max) = schema.get("maxLength").and_then(|m| m.as_u64()) {
            if (text.chars().count() as u64) > max {
                return Some(format!("{path}: string length {} above maxLength {max}", text.chars().count()));
            }
        }
        if let Some(pattern) = schema.get("pattern").and_then(|p| p.as_str()) {
            match Regex::new(pattern) {
                Ok(re) => {
                    if !re.is_match(text) {
                        return Some(format!("{path}: string does not match configured pattern"));
                    }
                }
                Err(_) => return Some(format!("{path}: invalid or unsafe regex pattern")),
            }
        }
    }

    if let (Some(arr), Some(items_schema)) = (node.as_array(), schema.get("items")) {
        for (i, item) in arr.iter().enumerate() {
            let item_path = format!("{path}[{i}]");
            if let Some(err) = validate_node(item, items_schema, &item_path) {
                return Some(err);
            }
        }
    }

    None
}

fn check_type(node: &Value, expected: &str, path: &str) -> Option<String> {
    let ok = match expected {
        "null" => node.is_null(),
        "string" => node.is_string(),
        "number" => node.is_number(),
        "integer" => {
            node.is_i64()
                || node.is_u64()
                || node.as_f64().map(|f| f.fract() == 0.0).unwrap_or(false)
        }
        "boolean" => node.is_boolean(),
        "object" => node.is_object(),
        "array" => node.is_array(),
        _ => true,
    };
    if ok {
        None
    } else {
        Some(format!("{path}: expected {expected} but got a different type"))
    }
}

impl Plugin for JsonSchemaValidator {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        Ok(Some(Self::validate(payload, config)?))
    }

    fn execute_with_result(&self, payload: &[u8], config: &str) -> Result<PluginResult, PluginError> {
        let text = std::str::from_utf8(payload)?;
        let (schema_config, on_invalid) = parse_config(config);
        let idempotency_key = self.idempotency_key(payload, config);
        match validate_payload(text, &schema_config)? {
            None => Ok(PluginResult::success(Some(payload.to_vec()), idempotency_key)),
            Some(err) => Ok(match on_invalid {
                OnInvalid::Skip => PluginResult::skip(idempotency_key, err),
                OnInvalid::Fail => PluginResult::fail(idempotency_key, err),
                OnInvalid::Dlq => PluginResult::dlq(Some(payload.to_vec()), idempotency_key, err),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::result::ErrorStrategy;

    #[test]
    fn compact_required_passes_when_present() {
        let out = JsonSchemaValidator::validate(
            r#"{"order_id":1,"amount":9.5,"customer_email":"a@x.com"}"#,
            "required=order_id,amount,customer_email",
        );
        assert!(out.is_ok());
    }

    #[test]
    fn compact_required_reports_first_missing() {
        let err = JsonSchemaValidator::validate(
            r#"{"_value":"default"}"#,
            "required=order_id,amount,customer_email",
        )
        .unwrap_err();
        assert_eq!(err.to_string(), "$: missing required field 'order_id'");
    }

    #[test]
    fn full_schema_type_and_required() {
        let schema = r#"{"type":"object","required":["id"],"properties":{"id":{"type":"integer"}}}"#;
        assert!(JsonSchemaValidator::validate(r#"{"id":5}"#, schema).is_ok());
        assert!(JsonSchemaValidator::validate(r#"{"id":"x"}"#, schema).is_err());
        assert!(JsonSchemaValidator::validate(r#"{}"#, schema).is_err());
    }

    #[test]
    fn invalid_routes_to_dlq_by_default() {
        let schema = r#"{"type":"object","required":["name"]}"#;
        let plugin = JsonSchemaValidator::new();
        let result = plugin
            .execute_with_result(br#"{"email":"a@b.com"}"#, schema)
            .unwrap();
        assert_eq!(result.error_strategy(), Some(ErrorStrategy::Dlq));
    }

    #[test]
    fn on_invalid_skip_and_fail() {
        let plugin = JsonSchemaValidator::new();
        let skip = plugin
            .execute_with_result(
                br#"{"email":"a@b.com"}"#,
                r#"schema={"type":"object","required":["name"]};onInvalid=skip"#,
            )
            .unwrap();
        assert_eq!(skip.error_strategy(), Some(ErrorStrategy::Skip));

        let fail = plugin
            .execute_with_result(
                br#"{"email":"a@b.com"}"#,
                r#"schema={"type":"object","required":["name"]};onInvalid=fail"#,
            )
            .unwrap();
        assert_eq!(fail.error_strategy(), Some(ErrorStrategy::Fail));
    }

    #[test]
    fn on_invalid_with_compact_config() {
        let plugin = JsonSchemaValidator::new();
        let invalid = plugin
            .execute_with_result(br#"{"order_id":7}"#, "required=order_id,amount;onInvalid=skip")
            .unwrap();
        assert_eq!(invalid.error_strategy(), Some(ErrorStrategy::Skip));

        let valid = plugin
            .execute_with_result(
                br#"{"order_id":7,"amount":12.5}"#,
                "required=order_id,amount;onInvalid=skip",
            )
            .unwrap();
        assert!(valid.is_success());
    }

    #[test]
    fn valid_payload_succeeds_with_on_invalid_present() {
        let plugin = JsonSchemaValidator::new();
        let result = plugin
            .execute_with_result(
                br#"{"name":"Alice"}"#,
                r#"schema={"type":"object","required":["name"]};onInvalid=fail"#,
            )
            .unwrap();
        assert!(result.is_success());
    }
}

File: durga-runtime/src/main/java/org/gautelis/durga/plugins/JsonSchemaValidator.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.micrometer.core.instrument.Timer;
import io.micrometer.core.instrument.Counter;
import org.gautelis.durga.monitoring.Metrics;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Validates JSON payloads against a JSON Schema-like configuration.
 *
 * <p>Supported validations:
 * <ul>
 * <li>{@code type} --- string, number, integer, boolean, object, array, null</li>
 * <li>{@code required} --- list of required field names</li>
 * <li>{@code properties} --- nested schema validation per field</li>
 * <li>{@code enum} --- whitelist of allowed values</li>
 * <li>{@code minimum} / {@code maximum} --- numeric range</li>
 * <li>{@code minLength} / {@code maxLength} --- string length bounds</li>
 * <li>{@code pattern} --- regex pattern (Java syntax)</li>
 * </ul>
 *
 * <p>The config is a JSON object matching the schema subset above.
 *
 * <p>An optional {@code onInvalid} directive selects how invalid payloads are
 * handled: {@code dlq} (default) routes them to the dead-letter channel,
 * {@code skip} drops them silently, and {@code fail} fails the process. The
 * directive is supplied alongside the schema, e.g.
 * {@code schema={"type":"object"};onInvalid=skip} or, in compact form,
 * {@code required=order_id,amount;onInvalid=fail}.
 */
public final class JsonSchemaValidator implements Plugin {

    private static final Logger LOG = LoggerFactory.getLogger(JsonSchemaValidator.class);

    private final ObjectMapper mapper = new ObjectMapper();
    private JsonNode schema;

    /** How the plugin handles a payload that fails validation. */
    enum OnInvalid {
        DLQ, SKIP, FAIL
    }

    record ValidatorConfig(String schemaConfig, OnInvalid onInvalid) {
    }

    @Override
    public String execute(String payload, String config) throws Exception {
        String pluginName = "json-schema-validator";
        Counter counter = Metrics.registry().counter("plugin.executions", "plugin", pluginName);
        Timer timer = Metrics.registry().timer("plugin.duration", "plugin", pluginName);
        counter.increment();
        return timer.recordCallable(() -> {
            String error = validatePayload(payload, parseConfig(config).schemaConfig());
            if (error != null) {
                throw new ValidationException(error);
            }
            return payload;
        });
    }

    @Override
    public PluginResult executeWithResult(byte[] payload, String config) throws Exception {
        String pluginName = "json-schema-validator";
        Counter counter = Metrics.registry().counter("plugin.executions", "plugin", pluginName);
        Timer timer = Metrics.registry().timer("plugin.duration", "plugin", pluginName);
        counter.increment();
        return timer.recordCallable(() -> {
            ValidatorConfig parsed = parseConfig(config);
            String error = validatePayload(Plugin.toString(payload), parsed.schemaConfig());
            String idempotencyKey = idempotencyKey(payload, config);
            if (error == null) {
                return PluginResult.success(payload, idempotencyKey);
            }
            return switch (parsed.onInvalid()) {
                case SKIP -> PluginResult.skip(idempotencyKey, error);
                case FAIL -> PluginResult.fail(idempotencyKey, error);
                case DLQ -> PluginResult.dlq(payload, idempotencyKey, error);
            };
        });
    }

    private String validatePayload(String payload, String schemaConfig) throws Exception {
        JsonNode input = mapper.readTree(payload);
        if (schemaConfig == null || schemaConfig.isBlank()) {
            // No schema configured: nothing to validate against, treat as valid.
            return null;
        }
        if (!schemaConfig.trim().startsWith("{")) {
            return validateCompactConfig(input, schemaConfig);
        }
        JsonNode schemaNode = mapper.readTree(schemaConfig);
        return validate(input, schemaNode, "$");
    }

    /**
     * Separates an optional {@code onInvalid} directive from the schema
     * configuration. Segments are split on top-level {@code ;} only, so a
     * {@code ;} inside the JSON schema (within braces or string literals) does
     * not act as a separator.
     */
    static ValidatorConfig parseConfig(String config) {
        if (config == null) {
            return new ValidatorConfig(null, OnInvalid.DLQ);
        }
        OnInvalid onInvalid = OnInvalid.DLQ;
        StringBuilder schemaConfig = new StringBuilder();
        for (String segment : splitTopLevel(config)) {
            String trimmed = segment.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            String lower = trimmed.toLowerCase(Locale.ROOT);
            if (lower.startsWith("oninvalid=")) {
                onInvalid = parseOnInvalid(trimmed.substring("onInvalid=".length()).trim());
            } else if (lower.startsWith("schema=")) {
                appendSchema(schemaConfig, trimmed.substring("schema=".length()));
            } else {
                appendSchema(schemaConfig, trimmed);
            }
        }
        return new ValidatorConfig(schemaConfig.toString(), onInvalid);
    }

    private static void appendSchema(StringBuilder schemaConfig, String segment) {
        String value = segment.trim();
        if (value.isEmpty()) {
            return;
        }
        if (schemaConfig.length() > 0) {
            schemaConfig.append(';');
        }
        schemaConfig.append(value);
    }

    private static OnInvalid parseOnInvalid(String value) {
        return switch (value.toLowerCase(Locale.ROOT)) {
            case "skip" -> OnInvalid.SKIP;
            case "fail" -> OnInvalid.FAIL;
            default -> OnInvalid.DLQ;
        };
    }

    private static List<String> splitTopLevel(String config) {
        List<String> parts = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        int depth = 0;
        boolean inString = false;
        boolean escaped = false;
        for (int i = 0; i < config.length(); i++) {
            char c = config.charAt(i);
            if (inString) {
                current.append(c);
                if (escaped) {
                    escaped = false;
                } else if (c == '\\') {
                    escaped = true;
                } else if (c == '"') {
                    inString = false;
                }
                continue;
            }
            switch (c) {
                case '"' -> {
                    inString = true;
                    current.append(c);
                }
                case '{', '[' -> {
                    depth++;
                    current.append(c);
                }
                case '}', ']' -> {
                    if (depth > 0) {
                        depth--;
                    }
                    current.append(c);
                }
                case ';' -> {
                    if (depth == 0) {
                        parts.add(current.toString());
                        current.setLength(0);
                    } else {
                        current.append(c);
                    }
                }
                default -> current.append(c);
            }
        }
        parts.add(current.toString());
        return parts;
    }

    private String validateCompactConfig(JsonNode input, String config) {
        for (String part : config.split(";")) {
            int eq = part.indexOf('=');
            if (eq <= 0) {
                continue;
            }
            String key = part.substring(0, eq).trim();
            String value = part.substring(eq + 1).trim();
            if ("required".equals(key)) {
                for (String field : value.split(",")) {
                    String path = field.trim();
                    if (!path.isEmpty() && PipelinePlugin.fieldAt(input, path) == null) {
                        return "$: missing required field '" + path + "'";
                    }
                }
            }
        }
        return null;
    }

    private String validate(JsonNode node, JsonNode schema, String path) {
        if (node == null || schema == null) {
            return null;
        }

        String type = textOrNull(schema, "type");
        if (type != null) {
            String typeErr = checkType(node, type, path);
            if (typeErr != null) return typeErr;
        }

        if (schema.has("enum")) {
            String enumErr = checkEnum(node, schema.get("enum"), path);
            if (enumErr != null) return enumErr;
        }

        if (type != null && type.equals("object") && node.isObject()) {
            if (schema.has("required")) {
                String reqErr = checkRequired(node, schema.get("required"), path);
                if (reqErr != null) return reqErr;
            }
            if (schema.has("properties")) {
                JsonNode props = schema.get("properties");
                for (Map.Entry<String, JsonNode> entry : props.properties()) {
                    String field = entry.getKey();
                    JsonNode childSchema = entry.getValue();
                    JsonNode child = node.get(field);
                    String childPath = path + "." + field;
                    if (child != null) {
                        String childErr = validate(child, childSchema, childPath);
                        if (childErr != null) return childErr;
                    }
                }
            }
        }

        if (node.isNumber()) {
            if (schema.has("minimum")) {
                if (node.asDouble() < schema.get("minimum").asDouble()) {
                    return path + ": numeric value below minimum " + schema.get("minimum");
                }
            }
            if (schema.has("maximum")) {
                if (node.asDouble() > schema.get("maximum").asDouble()) {
                    return path + ": numeric value above maximum " + schema.get("maximum");
                }
            }
        }

        if (node.isTextual()) {
            String text = node.asText();
            if (schema.has("minLength")) {
                if (text.length() < schema.get("minLength").asInt()) {
                    return path + ": string length " + text.length() + " below minLength " + schema.get("minLength").asInt();
                }
            }
            if (schema.has("maxLength")) {
                if (text.length() > schema.get("maxLength").asInt()) {
                    return path + ": string length " + text.length() + " above maxLength " + schema.get("maxLength").asInt();
                }
            }
            if (schema.has("pattern")) {
                String regex = schema.get("pattern").asText();
                try {
                    String boundedText = PipelinePlugin.requireBoundedRegexInput(text);
                    if (!PipelinePlugin.compileSafeRegex(regex).matcher(boundedText).matches()) {
                        return path + ": string does not match configured pattern";
                    }
                } catch (IllegalArgumentException e) {
                    return path + ": invalid or unsafe regex pattern";
                }
            }
        }

        if (node.isArray() && schema.has("items")) {
            JsonNode itemsSchema = schema.get("items");
            for (int i = 0; i < node.size(); i++) {
                String itemPath = path + "[" + i + "]";
                String itemErr = validate(node.get(i), itemsSchema, itemPath);
                if (itemErr != null) return itemErr;
            }
        }

        return null;
    }

    private static String checkType(JsonNode node, String expected, String path) {
        boolean ok = switch (expected) {
            case "null" -> node.isNull();
            case "string" -> node.isTextual();
            case "number" -> node.isNumber();
            case "integer" -> node.isInt() || node.isLong() || (node.isDouble() && node.asDouble() == Math.floor(node.asDouble()));
            case "boolean" -> node.isBoolean();
            case "object" -> node.isObject();
            case "array" -> node.isArray();
            default -> true;
        };
        return ok ? null : path + ": expected " + expected + " but got " + node.getNodeType();
    }

    private static String checkRequired(JsonNode node, JsonNode required, String path) {
        for (JsonNode field : required) {
            String name = field.asText();
            if (!node.has(name)) {
                return path + ": missing required field '" + name + "'";
            }
        }
        return null;
    }

    private static String checkEnum(JsonNode node, JsonNode allowed, String path) {
        for (JsonNode value : allowed) {
            if (node.equals(value)) {
                return null;
            }
        }
        return path + ": value not in allowed enum";
    }

    private static String textOrNull(JsonNode node, String field) {
        JsonNode value = node.get(field);
        return value != null ? value.asText() : null;
    }

    /** Thrown when validation fails; caught by the plugin executor's DLQ handler. */
    public static final class ValidationException extends Exception {
        public ValidationException(String message) {
            super(message);
        }
    }
}
